clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java

```diff
@@ -200,6 +200,9 @@ protected void handleFetchSuccess(final Node fetchTarget,
                     if (partitionData.currentLeader().leaderId() != -1 && partitionData.currentLeader().leaderEpoch() != -1) {
                         partitionsWithUpdatedLeaderInfo.put(partition, new Metadata.LeaderIdAndEpoch(
                             Optional.of(partitionData.currentLeader().leaderId()), Optional.of(partitionData.currentLeader().leaderEpoch())));
+                    } else {
+                        requestMetadataUpdate(metadata, subscriptions, partition);
+                        subscriptions.awaitUpdate(partition);
```

Comment:
With this change, we first request a metadata update, then flag our partition as awaiting the metadata update whenever we encounter a `NOT_LEADER_OR_FOLLOWER` or `FENCED_LEADER_EPOCH`. However, in the `FetchCollector.handleInitializeErrors()` method, we only only request the metadata update, but _don't_ flag the partition. Is that seeming inconsistency intentional?

Reply:
If the FetchStates is FETCHING as per KIP-951, the FetchCollector.handleInitializeErrors() method is called.
I thought that in this case, it should not be changed to AWAIT_UPDATE.
Additionally, if it's AWAIT_UPDATE, it will be filtered out by the following code inside the FetchCollector.initialize() method and will not go through FetchCollector.handleInitializeErrors().

```
if (!subscriptions.hasValidPosition(tp)) {
   // this can happen when a rebalance happened while fetch is still in-flight
   log.debug("Ignoring fetched records for partition {} since it no longer has valid position", tp);
   return null;
}
```

Reply:
As another alternative, it could change the status to AWAIT_UPDATE in FetchCollector.handleInitializeErrors() only when it's not a KIP-951 case.
Upon further thought, it seems possible to differentiate based on the following conditions.
```
completedFetch.partitionData.currentLeader().leaderId() != -1 && completedFetch.partitionData.currentLeader().leaderEpoch() != -1
```